### PR TITLE
Distributor limits

### PR DESCRIPTION
**Pull request by pstibrany**
**What this PR does**: This PR introduces two new limits for distributor: max ingestion rate (samples/sec) that distributor can accept, and max number of inflight requests. When these limits are reached, distributor returns error which is translated into 5xx.

In addition to limits, new metrics are introduced. These expose current values as well as limits.

~This PR builts on top of #4070, and will be rebased after #4070 is merged.~ Done.

**Checklist**
- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/distributor/distributor.go [hunk 1]
```diff
@@ -137,6 +151,8 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
 	f.BoolVar(&cfg.ShardByAllLabels, "distributor.shard-by-all-labels", false, "Distribute samples based on all labels, as opposed to solely by user and metric name.")
 	f.StringVar(&cfg.ShardingStrategy, "distributor.sharding-strategy", util.ShardingStrategyDefault, fmt.Sprintf("The sharding strategy to use. Supported values are: %s.", strings.Join(supportedShardingStrategies, ", ")))
 	f.BoolVar(&cfg.ExtendWrites, "distributor.extend-writes", true, "Try writing to an additional ingester in the presence of an ingester not in the ACTIVE state. It is useful to disable this along with -ingester.unregister-on-shutdown=false in order to not spread samples to extra ingesters during rolling restarts with consistent naming.")
+	f.Float64Var(&cfg.MaxIngestionRate, "distributor.max-ingestion-rate", 0, "Max ingestion rate (samples/sec) that distributor will accept. This limit is per-distributor, not per-tenant. Additional push requests will be rejected. Current ingestion rate is computed as exponentially weighted moving average, updated every second. 0 = unlimited.")
```
Comment (stevesg) [inline pkg/distributor/distributor.go:154]:
```suggestion
 	f.Float64Var(&cfg.MaxIngestionRate, "distributor.max-ingestion-rate", 0, "Max ingestion rate (samples/sec) that this distributor will accept. This limit is per-distributor, not per-tenant. Additional push requests will be rejected. Current ingestion rate is computed as exponentially weighted moving average, updated every second. 0 = unlimited.")
```

File: pkg/distributor/distributor_test.go [hunk 2]
```diff
@@ -446,6 +446,156 @@ func TestDistributor_PushIngestionRateLimiter(t *testing.T) {
...
+				numDistributors:     1,
+				shardByAllLabels:    true,
+				limits:              limits,
+				maxInflightRequests: testData.inflightLimit,
+				maxIngestionRate:    testData.ingestionRateLimit,
+			})
+			defer stopAll(distributors, r)
+
+			d := distributors[0]
+			d.inflightPushRequests.Add(int64(testData.preInflight))
```
Comment (stevesg) [inline pkg/distributor/distributor_test.go:574]: This means we're not testing that the `Inc`/`Dec` is correct, but I can't think of a simple alternative. Unless there was an easy way to make a `Push()` request block for a long time, maybe it's not worth the effort.

**Review by stevesg — APPROVED** [review]
LGTM - clear to understand.

File: pkg/distributor/distributor.go [hunk 3]
```diff
@@ -137,6 +151,8 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
 	f.BoolVar(&cfg.ShardByAllLabels, "distributor.shard-by-all-labels", false, "Distribute samples based on all labels, as opposed to solely by user and metric name.")
 	f.StringVar(&cfg.ShardingStrategy, "distributor.sharding-strategy", util.ShardingStrategyDefault, fmt.Sprintf("The sharding strategy to use. Supported values are: %s.", strings.Join(supportedShardingStrategies, ", ")))
 	f.BoolVar(&cfg.ExtendWrites, "distributor.extend-writes", true, "Try writing to an additional ingester in the presence of an ingester not in the ACTIVE state. It is useful to disable this along with -ingester.unregister-on-shutdown=false in order to not spread samples to extra ingesters during rolling restarts with consistent naming.")
+	f.Float64Var(&cfg.MaxIngestionRate, "distributor.max-ingestion-rate", 0, "Max ingestion rate (samples/sec) that this distributor will accept. This limit is per-distributor, not per-tenant. Additional push requests will be rejected. Current ingestion rate is computed as exponentially weighted moving average, updated every second. 0 = unlimited.")
```
Comment (pracucci) [inline pkg/distributor/distributor.go:154]: In the ingester the config is `-ingester.instance-limits.max-ingestion-rate` (has `instance-limits`). What do you think if we keep it specular in distributors too?

_Same comment applies to `max-inflight-push-requests` too._

File: pkg/distributor/distributor.go [hunk 4]
```diff
@@ -137,6 +151,8 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
...
+	f.IntVar(&cfg.MaxInflightPushRequests, "distributor.max-inflight-push-requests", 0, "Max inflight push requests that this distributor can handle. Additional requests will be rejected. 0 = unlimited.")
```
Comment (pracucci) [inline pkg/distributor/distributor.go:155]: I would mention "This limit is per-distributor, not per-tenant" here as well.

File: pkg/distributor/distributor.go [hunk 5]
```diff
@@ -273,6 +290,29 @@ func New(cfg Config, clientConfig ingester_client.Config, limits *validation.Ove
 			Help: "Unix timestamp of latest received sample per user.",
 		}, []string{"user"}),
 	}
+
+	promauto.With(reg).NewGaugeFunc(prometheus.GaugeOpts{
+		Name: "cortex_distributor_inflight_push_requests",
```
Comment (pracucci) [inline pkg/distributor/distributor.go:295]: In the ingester we have 1 metric exposing limits with the limit as label. Here we use a different pattern. I would keep the two consistent.

File: pkg/distributor/distributor.go [hunk 6]
```diff
@@ -603,6 +663,8 @@ func (d *Distributor) Push(ctx context.Context, req *cortexpb.WriteRequest) (*co
 		return nil, httpgrpc.Errorf(http.StatusTooManyRequests, "ingestion rate limit (%v) exceeded while adding %d samples and %d metadata", d.ingestionRateLimiter.Limit(now, userID), validatedSamples, len(validatedMetadata))
 	}
 
+	d.ingestionRate.Add(int64(validatedSamples))
```
Comment (pracucci) [inline pkg/distributor/distributor.go:666]: Should we use `totalN` instead of `validatedSamples` like we do for the per-tenant limiter?

**Review by pracucci — COMMENTED** [review]
Good job @pstibrany and thanks for working on distributor limits! I left few comments, overall LGTM!

File: pkg/distributor/distributor.go (hunk 6 above)
Comment (pstibrany) [reply pkg/distributor/distributor.go:666]: I was considering that. Problem is that in ingester, we don't include metadata in this computation. I think we should be consistent – either include metadata in both, or not.

Comment (pracucci) [reply pkg/distributor/distributor.go:666]: Good point. I agree we need to be consistent. Not including it, means there's no limit applied to metadata, so I could intentionally skip this limit with remote write requests including only metadata. Including metadata both in distributor and ingester may be safer.

File: pkg/distributor/distributor.go (hunk 3 above)
Comment (pstibrany) [reply pkg/distributor/distributor.go:154]: Do you suggest to introduce `instance_limits` field in YAML as well?

Comment (pstibrany) [reply pkg/distributor/distributor.go]: > Do you suggest to introduce instance_limits field in YAML as well?

I've done it that way.

**Comment by pstibrany** [comment]
Thank you for reviews! I've addressed your feedback, please take a look again.

File: pkg/distributor/distributor.go (hunk 3 above)
Comment (pracucci) [reply pkg/distributor/distributor.go]: Yes, to keep consistency between CLI and YAML.

**Review by pracucci — APPROVED** [review]
Very good job! LGTM (modulo the CHANGELOG fix). Thanks!

**Pull request merged**